### PR TITLE
BUGFIX: Fix support for nullable return type declarations

### PR DESCRIPTION
**Pull request by bwaidelich**
Fixes: #900

### CODE REVIEW

**Comment by bwaidelich** [comment]
This breaks functional tests with PHP < 7.1 because the compilation step fails (`syntax error, unexpected '?'`)

File: TYPO3.Flow/Classes/TYPO3/Flow/Object/Proxy/ProxyMethod.php [hunk 1]
```diff
@@ -155,10 +155,6 @@ public function render()
         $visibility = ($this->visibility === null ? $this->getMethodVisibilityString() : $this->visibility);
 
         $returnType = $this->reflectionService->getMethodDeclaredReturnType($this->fullOriginalClassName, $this->methodName);
-        // TODO: Revise "TypeHandling::isSimpleType" and check if we want to add null.
-        if ($returnType !== null && !TypeHandling::isSimpleType($returnType) && !in_array($returnType, ['self', 'null', 'callable'])) {
-            $returnType = '\\' . $returnType;
-        }
```
Comment (bwaidelich) [inline TYPO3.Flow/Classes/TYPO3/Flow/Object/Proxy/ProxyMethod.php]: I moved the logic to the ReflectionService to be in sync with related code

File: TYPO3.Flow/Tests/Functional/Aop/Fixtures/TargetClassWithPhp71Features.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * This file is part of the TYPO3.Flow package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+class TargetClassWithPhp71Features
```
Comment (bwaidelich) [inline TYPO3.Flow/Tests/Functional/Aop/Fixtures/TargetClassWithPhp71Features.php]: This is the culprit that breaks PHP < 7.1 support

File: TYPO3.Flow/Tests/Functional/Aop/Fixtures/TargetClassWithPhp7Features.php [hunk 3]
```diff
@@ -21,10 +21,28 @@ class TargetClassWithPhp7Features
...
-        return "{$aString} and {$aNumber}";
+        return "{$aString} and {$aNumber} and {$anObject}";
     }
+
+    public function methodWithStaticScalarReturnTypeDeclaration(): string
+    {
+        return 'it works';
+    }
+
+    public function methodWithStaticObjectReturnTypeDeclaration(): TargetClassWithPhp7Features
```
Comment (bwaidelich) [inline TYPO3.Flow/Tests/Functional/Aop/Fixtures/TargetClassWithPhp7Features.php]: This implicitly tests #967 as well

**Comment by bwaidelich** [comment]
I wonder if we should remove the tests in order to get this in.. At least I have no other idea on how we could solve the PHP < 7.1 issue..

**Comment by kitsunet** [comment]
I guess I would be ok with that :)

**Comment by albe** [comment]
Yes, fine with that. And let's increase PHP version requirement with Flow5

**Comment by bwaidelich** [comment]
@aertmann hehe. I tried everything I could think of for several hours (including evil eval) but nothing worked out, so I just went for the easy way out for now

File: TYPO3.Flow/Classes/TYPO3/Flow/Reflection/MethodReflection.php [hunk 4]
```diff
@@ -107,6 +107,18 @@ public function getDeclaredReturnType()
     }
 
     /**
+     * @return bool Whether a return type is declared (@see getDeclaredReturnType()) and has a nullable modified ("?")
```
Comment (kdambekalns) [inline TYPO3.Flow/Classes/TYPO3/Flow/Reflection/MethodReflection.php]: probably "a nullable modifier" was intended?

Comment (bwaidelich) [inline TYPO3.Flow/Classes/TYPO3/Flow/Reflection/MethodReflection.php]: Right, I'll push a follow-up and merge it. Thanks!

**Comment by bwaidelich** [comment]
Merging with previous votes (just rebased and fixed a typo afterwards)

**Pull request merged**